tempesta_fw/http.c

```diff
@@ -517,17 +562,19 @@ static void
 tfw_http_req_cache_cb(TfwHttpReq *req, TfwHttpResp *resp)
 {
 	int r;
-	TfwConnection *conn;
+	int req_conn_close = !!(req->flags & TFW_HTTP_CONN_CLOSE);
+	TfwConnection *srv_conn, *cli_conn = req->conn;
 
 	if (resp) {
 		/*
-		 * We have prepared response, send it as is.
-		 * The response is pass through ot was generated from
-		 * the cache, so unrefer all its data.
+		 * The response is prepared, send it as is. The response
+		 * is either passed through from the back-end server, or
+		 * it is generated from the cache, so unrefer all its data.
 		 */
-		if (tfw_http_adjust_resp(resp, req))
-			return;
-		tfw_connection_send(req->conn, (TfwMsg *)resp, true);
+		if (tfw_http_adjust_resp(resp, req) == 0)
+			tfw_connection_send(cli_conn, (TfwMsg *)resp, true);
+		if (req_conn_close)
+			tfw_connection_drop(cli_conn);
```

Comment:
Looks like there was a bug here after recent changes in cache processing. `req` and `resp` should be deleted here.